plan/gen_physical_plans.go

```diff
@@ -143,6 +144,46 @@ func (p *LogicalJoin) getMergeJoin(prop *requiredProp) []PhysicalPlan {
 	return joins
 }
 
+func (p *LogicalJoin) getEnforcedMergeJoin(prop *requiredProp) []PhysicalPlan {
+	// Check whether SMJ can satisfy the parents property
+	for _, col := range prop.cols {
+		isExist := false
+		for _, key := range p.LeftJoinKeys {
+			if reflect.DeepEqual(*key, *col) {
+				isExist = true
+				break
+			}
+		}
+		for _, key := range p.RightJoinKeys {
+			if reflect.DeepEqual(*key, *col) {
+				isExist = true
+				break
+			}
+		}
+		if !isExist && (prop == nil || !prop.enforced) {
```

Comment:
What happened if the sql is:
```sql
drop table if exists t1, t2;
create table t1(a bigint, b bigint);
create table t2(a bigint, b bigint);
select /*+ TIDB_SMJ(t1, t2) */ * from t1 join t2 on t1.a = t2.a order by t2.b, t2.a;
```

Reply:
```
tidb > explain select /*+ TIDB_SMJ(t1, t2) */ * from t1 join t2 on t1.a = t2.a order by t2.b, t2.a;
+----------------+-------------+-----------------+------+---------------------------------------------------------------------------------------+----------+
| id             | parents     | children        | task | operator info                                                                         | count    |
+----------------+-------------+-----------------+------+---------------------------------------------------------------------------------------+----------+
| TableScan_10   |             |                 | cop  | table:t1, range:[-inf,+inf], keep order:false                                         | 10000.00 |
| TableReader_11 | Sort_12     |                 | root | data:TableScan_10                                                                     | 10000.00 |
| Sort_12        | MergeJoin_9 | TableReader_11  | root | root.t1.a:asc                                                                         | 10000.00 |
| TableScan_13   |             |                 | cop  | table:t2, range:[-inf,+inf], keep order:false                                         | 10000.00 |
| TableReader_14 | Sort_15     |                 | root | data:TableScan_13                                                                     | 10000.00 |
| Sort_15        | MergeJoin_9 | TableReader_14  | root | root.t2.a:asc                                                                         | 10000.00 |
| MergeJoin_9    | Sort_6      | Sort_12,Sort_15 | root | inner join, equal:[eq(root.t1.a, root.t2.a)], left key:root.t1.a, right key:root.t2.a | 12500.00 |
| Sort_6         |             | MergeJoin_9     | root | root.t2.b:asc, root.t2.a:asc                                                          | 12500.00 |
+----------------+-------------+-----------------+------+---------------------------------------------------------------------------------------+----------+
8 rows in set (0.01 sec)
```

Reply:
This case is more specific:
```sql
TiDB(localhost:4000) > desc select /*+ TIDB_SMJ(t1, t2) */ * from t1 join t2 on t1.a = t2.a and t1.b = t2.b order by t2.b;
+----------------+--------------+-----------------+------+--------------------------------------------------------------------------------------------------------------------------------------+----------+
| id             | parents      | children        | task | operator info                                                                                                                        | count    |
+----------------+--------------+-----------------+------+--------------------------------------------------------------------------------------------------------------------------------------+----------+
| TableScan_10   |              |                 | cop  | table:t1, range:[-inf,+inf], keep order:false                                                                                        | 10000.00 |
| TableReader_11 | Sort_12      |                 | root | data:TableScan_10                                                                                                                    | 10000.00 |
| Sort_12        | MergeJoin_17 | TableReader_11  | root | test.t1.a:asc, test.t1.b:asc                                                                                                         | 10000.00 |
| TableScan_13   |              |                 | cop  | table:t2, range:[-inf,+inf], keep order:false                                                                                        | 10000.00 |
| TableReader_14 | Sort_15      |                 | root | data:TableScan_13                                                                                                                    | 10000.00 |
| Sort_15        | MergeJoin_17 | TableReader_14  | root | test.t2.a:asc, test.t2.b:asc                                                                                                         | 10000.00 |
| MergeJoin_17   |              | Sort_12,Sort_15 | root | inner join, equal:[eq(test.t1.a, test.t2.a) eq(test.t1.b, test.t2.b)], left key:test.t1.a, test.t1.b, right key:test.t2.a, test.t2.b | 12500.00 |
+----------------+--------------+-----------------+------+--------------------------------------------------------------------------------------------------------------------------------------+----------+
7 rows in set (0.00 sec)
```

The problem is, this enforced merge join can only order its output rows by left join keys **or** right join keys, if the left join keys are (t1.a, t1.b), right join keys are (t2.a, t2.b) but the required sort property on merge join's output is (t2.b), this enforced merge join can not satisfy this physical property, but we removed the sort upon it.

To fix this, we can:
1. check whether the required sort keys can be covered by the left or right join keys
2. YES: enforce the child to be sorted this way: (sort keys that in join keys, the rest join keys)
3. NO: return a `nil` PhysicalPlan

In this example, the child sort key can be: 
1. left: (t1.b, t1.a) 
2. right (t2.b, t2.a)

Then we can safely remove the sort upon the merge join.